ttnn::typecast row-major support

vtsilytskyiTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/16270

### Problem description
ROW-MAJOR typecast is not supported on device side.

### What's changed
Main change - add device-side typecast for ROW-MAJOR tensor. It was added in 2 places:

1. TypecastProgramFactory, which should theoretically handle short rows (which fits L1). Unfortunately, there is a bug in unary eltwise kernels, and rows >1024 elements are calculated incorrectly. So this execution path for RM is temporary disabled.
2. A new TypecastRowMajorChunkedProgramFactory, which takes any 32-aligned row, split it in chunks and compute in streaming manner. This implementation is correct and handle large inputs, but may be slower comparing to naive implementation above.

Additional changes:

1. Fix hashing in unary eltwise. shape.volume was enough for TILE layout, but for ROW_MAJOR it causes problems. Consider 32x64 kernel and then 64x32. Hashes will be same, but row size is different. Single test execution succeed, but sequence of tests will fail.

Generic improvements:

1. Initialize fields in `shared_variables_t` structure;
3. Use Tensor copy instead of const reference in `tensor_args_t`
4. Speedup unary eltwise write kernels. Not visible on single page reads, but should be noticeable on bigger workloads.

### Checklist

- [x] [![All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml/badge.svg?branch=vlad/row-major-for-typecast)](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:vlad/row-major-for-typecast)
- [x] [![Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml/badge.svg?branch=vlad/row-major-for-typecast)](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:vlad/row-major-for-typecast)
- [x] [![cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml/badge.svg?branch=vlad/row-major-for-typecast)](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:vlad/row-major-for-typecast)
- [x] New/Existing tests provide coverage for changes


#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [![(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml/badge.svg?branch=vlad/row-major-for-typecast)](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:vlad/row-major-for-typecast)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [![(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml/badge.svg?branch=vlad/row-major-for-typecast)](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:vlad/row-major-for-typecast)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [![(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml/badge.svg?branch=vlad/row-major-for-typecast)](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:vlad/row-major-for-typecast)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-model-perf-tests.yaml) tests)
  - [x] other selection - specify runs

vtsilytskyiTT commented:
Rebase, address copilot comments.

bklockiewiczTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,158 @@ def test_typecast_community(device):
+def run_typecast_row_major_test(shape, memory_config, input_dtype, output_dtype, device):
+    """Test typecast operation on row-major device tensors."""
+    torch.manual_seed(54321)
+    torch_dtype = torch.bfloat16
```
Is it the only one supported?

iwroszTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,158 @@ def test_typecast_community(device):
+    "shape",
+    [
+        (32, 64),  # 2D
+        (1, 32, 64),  # 3D
```
Could you also add some uneven shapes, e.g., (45, 70) (1, 17, 123)
and some with leading dimensions other than 1, e.g., (4, 32, 64), (8, 2, 32, 64)

iwroszTT commented on ttnn/cpp/ttnn/operations/copy/typecast/device/kernels/compute/eltwise_typecast_row_major.cpp:
```diff
@@ -0,0 +1,46 @@
+
+namespace NAMESPACE {
+void MAIN {
+    uint32_t per_core_block_cnt = get_compile_time_arg_val(0);
```
Compile time args can be constexpr

iwroszTT commented on ttnn/cpp/ttnn/operations/copy/typecast/device/kernels/dataflow/reader_typecast_row_major.cpp:
```diff
@@ -0,0 +1,29 @@
+        uint32_t l1_write_addr = get_write_ptr(cb_id_in0);
+        uint64_t src_noc_addr = get_noc_addr(page_id, s);
+        noc_async_read(src_noc_addr, l1_write_addr, page_size);
+        noc_async_read_barrier();
```
A note on performance, currently we read one page at a time, in other words, there will be max one page in-flight. For a small page_size reader perf will be limited by noc read latency.
The same applies to the writer kernel.

vtsilytskyiTT replied:
I'll reuse eltwise kernels for dataflow.

bklockiewiczTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,158 @@ def test_typecast_community(device):
+    input = torch.randn(shape, dtype=torch_dtype)
+
+    # Create row-major tensor on device
+    input_tensor = ttnn.from_torch(input, input_dtype, layout=ttnn.ROW_MAJOR_LAYOUT, device=device)
```
nit: is there a way to not rely on coversion here but define the tensor directly?

bklockiewiczTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,158 @@ def test_typecast_community(device):
+        dtype_map = {
+            ttnn.uint32: torch.int32,  # uint32 maps to int32 in torch
+        }
+        torch_output_dtype = dtype_map.get(output_dtype, torch.bfloat16)
```
Why not simply 
```python
if torch_output_dtype is None:
    # Fallback for dtypes not in the mapping (e.g., uint32)
    torch_output_dtype = torch.int32 if output_dtype == ttnn.uint32 else torch.bfloat16
```

vtsilytskyiTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,158 @@ def test_typecast_community(device):
+        "BFLOAT16_TO_UINT8",
+        "UINT8_TO_BFLOAT16",
+    ],
+)
```
Done

vtsilytskyiTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,158 @@ def test_typecast_community(device):
+        torch_expected = torch.clamp(torch_expected, min=0)
+
+    pcc_threshold = 0.99 if output_dtype == ttnn.bfloat8_b or input_dtype == ttnn.bfloat8_b else 0.9999
+    assert_with_pcc(torch_expected, torch_output, pcc_threshold)
```
For integers I'll use torch.equal, for FP assert_with_ulp.

vtsilytskyiTT committed:
Add typecast row-major support

Fix hashing for unary eltwise on the way.
Use noc_async_write_barrier to speedup the kernel.

vtsilytskyiTT commented:
Ready for review. Rebase, tests pass locally, update PR description. Run CI tests.

vtsilytskyiTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,191 @@ def test_typecast_community(device):
+    torch_output_dtype = get_ttnn_torch_dtype(output_dtype)
+    if torch_output_dtype is None:
+        # Fallback for dtypes not in the mapping (e.g., uint32)
+        torch_output_dtype = torch.int32 if output_dtype == ttnn.uint32 else torch.bfloat16
```
I'll just remove uint32 as useless as we already have int32.

vtsilytskyiTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,191 @@ def test_typecast_community(device):
+        # For integer types, use exact equality (typecast from float to int may have rounding)
+        assert torch.equal(
+            torch_expected, torch_output
+        ), f"Integer typecast mismatch: expected {torch_expected}, got {torch_output}"
```
False positive. It actually asserts, verified manually.

vtsilytskyiTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,191 @@ def test_typecast_community(device):
+    if is_integer_type:
+        assert torch.equal(
+            torch_output_rm, torch_output_tile
+        ), "Row-major and tile layouts should produce identical integer results"
```
False positive. It actually asserts, verified manually.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/copy/typecast/device/typecast_program_factory.cpp:
```diff
@@ -18,44 +19,52 @@ TypecastProgramFactory::cached_program_t TypecastProgramFactory::create(
-    uint32_t num_input_tiles = 2;
+    // Set CB page size correctly based on layout
+    // - For TILE layout: page = one 32x32 tile
+    // - For ROW_MAJOR layout: page = one full row including padding
```
for row major the recommended page size is 4kb. 
I recommend to cap it at that to avoid issues when working with tensors like [1, 10000000]

ayerofieiev-tt reviewed (approved):
my only concern is the switch from volume to shape in hashing

vtsilytskyiTT commented:
Address most of the comments, rebase.

dchenTT commented on ttnn/cpp/ttnn/operations/copy/typecast/device/typecast_device_op.cpp:
```diff
@@ -115,9 +108,27 @@ tt::stl::hash::hash_t TypecastDeviceOperation::compute_program_hash(
+            input_tensor.memory_config(),
+            input_shape,
+            input_tensor.layout());
+    }
```
The original hashing of using tensor volume is overly aggressive and inefficient. It should and will be improved not using that and there is an already issue filed for it. Now it is the same with using shape. It may be OK to leave like this in this PR but it will be changed later.

vtsilytskyiTT committed:
Add support for large RM tensors in typecast

mateusznowakTT commented on ttnn/cpp/ttnn/operations/copy/typecast/device/typecast_device_op.cpp:
```diff
@@ -80,8 +110,10 @@ void TypecastDeviceOperation::validate_on_program_cache_miss(
-                "Typecast operation requires output tensor to be in Tile layout when working with non-sharded tensor.");
+                preallocated_output_tensor.value().layout() == input_tensor.layout(),
+                "Typecast operation requires input and output layouts to match. Input layout: {}, Output layout: {}",
+                static_cast<int>(input_tensor.layout()),
```
You could use magic enum https://github.com/Neargye/magic_enum to have this print readable.

vtsilytskyiTT commented on tests/ttnn/unit_tests/operations/data_movement/test_copy_ops.py:
```diff
@@ -267,3 +268,190 @@ def test_typecast_community(device):
+        ), f"Integer typecast mismatch: expected {torch_expected}, got {torch_output}"
+    elif output_dtype == ttnn.bfloat8_b or input_dtype == ttnn.bfloat8_b:
+        # Use tighter tolerances for bfloat8_b due to lower precision
+        assert_with_ulp(torch_expected, torch_output, ulp_threshold=5)
```
I'm not testing bfloat8, so just removing it.

vtsilytskyiTT replied:
And removed unused integer types. I need only `int32`.

vtsilytskyiTT committed:
Address code review comments

vtsilytskyiTT committed:
Add more verification

vtsilytskyiTT commented:
/codeowners ping

vtsilytskyiTT committed:
Move copy test to nightly L2, change namespaces
